### ISSUE TITLE
disable screen blanking during system upgrade in text mode

### ISSUE THREAD

**Issue by kparal** [rpm-software-management/dnf-plugin-system-upgrade#26]
If you don't have plymouth or switch from graphical splash to text mode during system-upgrade, the screen will blank in 10 minutes (default timeout for screen blanking). Quite interestingly, it will not turn blank right away, but columns of characters gradually disappear before the screen is black.

Either way, it's a good idea to disable screen blanking during system upgrade, the same way we disable it during system installation. Otherwise some people might get confused and harm the upgrade process unintentionally. Please call:

```
setterm -blank 0
```

at the beginning of the upgrade process. That should disable screen blanking. Thanks.

**Comment by keszybz** [comment]
>  it will not turn blank right away, but columns of characters gradually disappear before the screen is black

Kernel bug.

> setterm -blank 0

Yes, this seem like a good idea. It could be implemented as a call to setterm, or through a direct write of the necessary string ("\33[9;0]") to the console. I checked util-linux sources and this is more or less what it does:

```
    /* See if the terminal is a virtual console terminal. */
    ctl->vcterm = (!strncmp(ctl->opt_te_terminal_name, "con", 3) ||
               !strncmp(ctl->opt_te_terminal_name, "linux", 5));
    if (!ctl->vcterm) {
        if (err)
            warnx(_("terminal %s does not support %s"),
                  ctl->opt_te_terminal_name, err);
    }

    printf("\033[9;%d]", ctl->opt_bl_min);
    return;
```

So I guess we could do it inline...

```
def enable_blanking(arg=10):
      sys.stdout.write("\33[9;{}]".format(arg))
def maybe_disable_blanking():
      term = os.getenv('TERM', '')
      if not ( term.startswith('linux') or term.startswith('con')):
             return
      atexit.register(enable_blanking)
      enable_blanking(0)
```

This escape sequence does not seem to have any effect on non-console terminals, so it all seems pretty safe.

**Comment by wgwoods** [comment]
> It could be implemented as a call to setterm, or through a direct write of the necessary string ("\33[9;0]") to the console.

I tried this in fedup, and never got it to work. The main problem seems to be that stdout isn't actually attached to the tty.

In 0.7.x (master) the upgrade is run with the default setting for `StandardOutput=`, so output just goes to the journal. In 0.5.x we have `StandardOutput=journal+console` but AFAIK that just means that systemd _redirects_ our output to the console, and I have no idea if it passes the control codes along as-is, or whether it's writing to `/dev/console` or `/dev/tty1`, and I'm pretty sure you need it to write to the actual tty for this to work.

Also - on _most_ (but not all!) systems, `plymouthd` is what's actually attached to tty1, and it may (or may not!) be twiddling/resetting the screen blank setting (and other VT settings) when you switch between the graphical plugin and the details plugin.

So in short: yes, in _theory_ it's easy to disable the screen blanking to work around this bug. If you can find a simple, reliable way to implement this workaround, send a PR, I guess, but even then: shouldn't this go upstream, in systemd or plymouth? Why make every long-running tool reimplement the workaround itself?

**Comment by wgwoods** [comment]
Let's also note, for the record, that this is all just to _work around_ an existing kernel bug, which only manifests after 10 minutes of inactivity _if_ you also aren't using the graphical boot splash, and you can press _literally any key_ and the screen fixes itself.

From what I've gathered from user reports, most real-world users _do_ have graphical boot splash, and they don't bother to sit and watch the upgrade. If they do come back and the screen is weird, even if their first instinct _is_ to hit Ctrl-Alt-Del, the screen fixes itself as soon as they hit Ctrl and/or Alt.

Testers like @kparal (and me) encounter this problem _far_ more often than it happens in the real world because: a) we watch the upgrade very closely, and b) we're running them in virtual machines that might not have proper graphical boot splash, and c) hitting "Ctrl+Alt" doesn't do anything in virtual machines.

IMHO The best solutions to this problem would be (in order):
1. Fix the kernel bug.
2. Fix the graphical splash for default virt guests
3. Find a workaround for the problem and add it to systemd/plymouthd (whichever is appropriate)
4. Document the bug and remind users/testers that they can hit, say, the Shift key to make the screen refresh itself
5. Work around the problem in any tool which encounters it.

I really don't like maintaining workarounds for someone else's bugs, especially just to avoid a problem that most people never even see, which can be fixed by just _not_ immediately yanking the power cord at the first sign of trouble. But I guess I could be convinced otherwise.

**Comment by kparal** [comment]
What's the kernel bug? I assumed this is a feature (console screen saver) and when playing with it in a standard installed system, it works as expected - changing the value changes the screen blanking timeout:
http://unix.stackexchange.com/questions/8056/disable-screen-blanking-on-text-console/32043#32043

The issue I know about is that it sometimes (with scrolling text, it seems) blanks just certain columns. But that seems just cosmetic, and we don't want it enabled anyway. There are certain types of applications which should disable it, I believe. Those are serious system interactions like system installation and system upgrade. Anaconda does it.

This blanking is causing us issues in test automation using OpenQA (which captures screenshots regularly and determines what to do next based on them). We had to use `setterm -blank 0` before running `dnf system-upgrade download`, so that we see properly when the download is finished (we run it in a text VT). That's fine by us, we have to do it for all long-running commands. But we would appreciate having the same experience during the upgrade as well, not just seeing a black screen after 10 minutes. But obviously we can't run the command manually at that phase. And we could work around the issue by sending e.g. Ctrl key periodically, but it introduces hacks into the framework and it would be better if it just worked out of the box. So that's our current motivation :)

Will, it didn't work for fedup, but maybe it might work now with systemd's offline upgrades environment? It would be great if you could give it a try (or we can try to hack it into it, but we don't have your experience in this area).

**Comment by keszybz** [comment]
> What's the kernel bug?

That some columns are not blanked properly resulting in garbled screen.

>  It would be great if you could give it a try

I'll try to put something together. It's more complicated than I thought because (as Will said) stdout is not attached to the terminal.

**Comment by keszybz** [comment]
Here we go, please test this if you get a chance...

**Comment by kparal** [comment]
Thanks for the patch! I can't make it today, but I'll definitely test #28 next week.

**Comment by keszybz** [comment]
I was finally able to do some proper testing of this patch. In short, it doesn't seem to break anything, either with or without plymouth running. Screen blanking seems to be successfully disabled. I was doing the tests in a VM under libvirt, and had various problems with the screen being garbled (I tried cirrus, QXL, VGA as the card types), but this happens the same with or without the blanking patch. The patch does not seem to break anything.
